### PR TITLE
Dropout

### PR DESCRIPTION
**Pull request by goodfeli**

### CODE REVIEW

File: cleverhans/picklable_model.py [hunk 1]
```diff
@@ -496,5 +506,36 @@ def set_input_shape(self, shape):
...
+
+    def fprop(self, x, dropout_dict=None, **kwargs):
+        """
+        Forward propagation as either no-op or dropping random units.
+        :param x: The input to the layer
+        :param dropout_dict: dict mapping layer names to dropout inclusion
+            probabilities.
+            This dictionary should be passed as a named argument to the MLP
+            class, which will then pass it to *all* layers' fprop methods.
+            Other layers will just recieve this as an ignored kwargs entry.
```
Comment (npapernot) [inline cleverhans/picklable_model.py]: Typo

File: cleverhans/picklable_model.py [hunk 2]
```diff
@@ -496,5 +506,36 @@ def set_input_shape(self, shape):
...
+        self.output_shape = shape
+
+    def get_params(self):
+        return []
+
+    def fprop(self, x, dropout_dict=None, **kwargs):
+        """
+        Forward propagation as either no-op or dropping random units.
+        :param x: The input to the layer
+        :param dropout_dict: dict mapping layer names to dropout inclusion
```
Comment (npapernot) [inline cleverhans/picklable_model.py]: I am curious what is the reason for using a dictionary common to the model rather than defining the dropout inclusion probabilities at each dropout layer individually?

Comment (goodfeli) [reply cleverhans/picklable_model.py]: I think i will rewrite it to define the inclusion probability for each layer in that layer's constructor, and then use fprop's kwargs only to say whether to drop or not drop. Does that sound good to you?

Comment (npapernot) [reply cleverhans/picklable_model.py]: Yes, that sounds easier to keep track of. Thanks!

File: cleverhans/picklable_model.py (hunk 1 above)
Comment (goodfeli) [reply cleverhans/picklable_model.py]: Fixed, thanks

File: cleverhans/picklable_model.py (hunk 2 above)
Comment (goodfeli) [reply cleverhans/picklable_model.py]: Done

**Comment by goodfeli** [comment]
I left the `dropout_dict` feature in so users can still change the dropout include_prob on each fprop if they need to (e.g. if it helps to answer a research question). Most users can ignore it and just use the `dropout` bool though.

**Comment by goodfeli** [comment]
A pytorch test failed. It seems unlikely that this PR affects the pytorch tests, so I think it was just a flaky test. I restarted that worker.

**Comment by goodfeli** [comment]
All tests pass, ready to re-review

File: cleverhans/picklable_model.py [hunk 3]
```diff
@@ -496,5 +509,48 @@ def set_input_shape(self, shape):
...
+            dropout include probability.
+            This dictionary should be passed as a named argument to the MLP
+            class, which will then pass it to *all* layers' fprop methods.
+            Other layers will just receive this as an ignored kwargs entry.
+            Each dropout layer looks up its own name in this dictionary
+            to read out its include probability.
+        """
+        include_prob = self.include_prob
+        if dropout_dict is not None:
+            assert dropout
```
Comment (npapernot) [inline cleverhans/picklable_model.py]: Just making sure I understand the intended use here. Does that mean that all layers must have `dropout` set to `True` even if their `include_prob` is 0 when `dropout_dict` is not `None`?  Does that also mean that an include probability must be specified for all dropout layers in the `dropout_dict`?

Comment (goodfeli) [reply cleverhans/picklable_model.py]: The default use case is that you never specify `include_prob` anywhere. During evaluation, you don't do anything special regarding dropout, and nothing gets dropped. During training, you pass "dropout=True" to make units get randomly dropped. If you've done nothing else, include_prob defaults to 0.5 in the Dropout class constructor.

A slightly more advanced use case is that you want to use different include_probs for some layer. For example, people usually use include_prob=0.8 for the input layer. To do this, you specify include_prob in the constructor arguments for those layers. Other than that, it's the same as the basic use case case. You do nothing special at test time and nothing is dropped. You pass dropout=True at train time and units are dropped based on their input prob.

The most advanced use case is if you want to change dropout include probabilities for a specific fprop call. In this case, you pass dropout=True and dropout_dict for that call. Each layer uses the include_prob specified in the dropout_dict for that call.

Comment (npapernot) [reply cleverhans/picklable_model.py]: My comment was mostly wondering about 2. and is more clear now, thanks for making the change! Also there was a typo in my comment, I meant when `include_prob` is 1. (basically disable dropout for one layer only), sorry about that.

**Comment by npapernot** [comment]
LGTM
